### PR TITLE
feat(hotkey): capture data part2 - declare coarse collector 

### PR DESCRIPTION
**Pull request by Smityz**
the implement of coarse collector 
## Config changes
```diff
[pegasus.server]
+ coarse_data_variance_threshold = 3
+ data_capture_hash_bucket_num = 37
```

### CODE REVIEW

File: src/server/hotkey_collector.h [hunk 1]
```diff
@@ -26,6 +26,12 @@ namespace server {
 
 class internal_collector_base;
 
+struct detect_hotkey_result
+{
+    int coarse_bucket_index;
+    detect_hotkey_result() : coarse_bucket_index(-1) {}
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:32]:
```suggestion
    int coarse_bucket_index = -1;
```

File: src/server/hotkey_collector.h [hunk 2]
```diff
@@ -75,28 +81,48 @@ class hotkey_collector : public dsn::replication::replica_base
     void analyse_data();
     void handle_rpc(const dsn::replication::detect_hotkey_request &req,
                     /*out*/ dsn::replication::detect_hotkey_response &resp);
+    static int get_bucket_id(dsn::string_view data);
```
Comment (levy5307) [inline src/server/hotkey_collector.h:84]: you shouldn't make this function to be a member of class `hotkey_collector`

File: src/server/hotkey_collector.cpp [hunk 3]
```diff
@@ -113,5 +144,63 @@ void hotkey_collector::on_stop_detect(dsn::replication::detect_hotkey_response &
...
+    double table_captured_key_sum = 0;
+    int hot_index = 0;
+    int hot_value = 0;
+    for (int i = 0; i < data_size; i++) {
+        table_captured_key_sum += captured_keys[i];
+        if (captured_keys[i] > hot_value) {
+            hot_index = i;
+            hot_value = captured_keys[i];
+        }
+    }
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:187]: add a blank line

File: src/server/hotkey_collector.h [hunk 4]
```diff
@@ -75,28 +81,48 @@ class hotkey_collector : public dsn::replication::replica_base
...
+    virtual void analyse_data(detect_hotkey_result &result) = 0;
 };
 
 // used in hotkey_collector_state::STOPPED and hotkey_collector_state::FINISHED, avoid null pointers
 class hotkey_empty_data_collector : public internal_collector_base
 {
 public:
-    void capture_data(const dsn::blob &hash_key, uint64_t size) {}
-    void analyse_data() {}
+    explicit hotkey_empty_data_collector(replica_base *base) : internal_collector_base(base){};
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:108]:
```suggestion
    explicit hotkey_empty_data_collector(replica_base *base) : internal_collector_base(base) {}
```

File: src/server/hotkey_collector.cpp [hunk 5]
```diff
@@ -113,5 +144,63 @@ void hotkey_collector::on_stop_detect(dsn::replication::detect_hotkey_response &
...
+    double standard_deviation = 0;
+    for (int i = 0; i < data_size; i++) {
+        if (i != hot_index) {
+            standard_deviation += pow((captured_keys[i] - captured_keys_avg_count), 2);
+        }
+    }
+    standard_deviation = sqrt(standard_deviation / (data_size - 2));
+    double hot_point = (hot_value - captured_keys_avg_count) / standard_deviation;
+    detect_hotkey_result result;
+    hot_point > threshold ? result.coarse_bucket_index = hot_index
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:200]:
```
if (hot_point > threshold) {
     result.coarse_bucket_index = hot_index;
}
```

File: src/server/hotkey_collector.cpp [hunk 6]
```diff
@@ -113,5 +144,63 @@ void hotkey_collector::on_stop_detect(dsn::replication::detect_hotkey_response &
...
+        _hash_buckets[i].store(0);
+    }
+    result = internal_analysis_method(buckets, FLAGS_coarse_data_variance_threshold);
+}
+
+detect_hotkey_result
+hotkey_coarse_data_collector::internal_analysis_method(const std::vector<uint64_t> &captured_keys,
+                                                       int threshold)
+{
+
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:174]:
```suggestion
```

File: src/server/hotkey_collector.cpp [hunk 7]
```diff
@@ -113,5 +144,63 @@ void hotkey_collector::on_stop_detect(dsn::replication::detect_hotkey_response &
...
+    int data_size = captured_keys.size();
+    dassert(captured_keys.size() > 2, "data_capture_hash_bucket_num is too small");
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:176]:
```suggestion
    dcheck_gt(captured_keys.size(), 2);
```

File: src/server/hotkey_collector.h (hunk 2 above)
Comment (Smityz) [reply src/server/hotkey_collector.h:84]: it will be used in `hotkey_coarse_data_collector` and `hotkey_fine_data_collector`
I think it is a public method

**Review by acelyc111 — APPROVED** [review]

**Comment by acelyc111** [comment]
> the implement of coarse collector
> 
> ## Config changes
> ```diff
> [pegasus.server]
> + coarse_data_variance_threshold
> + data_capture_hash_bucket_num
> ```
Better to add the default value in PR description.

**Review by acelyc111 — DISMISSED** [review]

File: src/server/hotkey_collector.cpp [hunk 8]
```diff
@@ -113,5 +144,64 @@ void hotkey_collector::on_stop_detect(dsn::replication::detect_hotkey_response &
     ddebug_replica(hint);
 }
 
+hotkey_coarse_data_collector::hotkey_coarse_data_collector(replica_base *base)
+    : internal_collector_base(base), _hash_buckets(FLAGS_data_capture_hash_bucket_num)
+{
+    for (std::atomic<uint64_t> &bucket : _hash_buckets) {
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:150]: `for (auto &bucket: _hash_buckets)`

File: src/server/hotkey_collector.cpp [hunk 9]
```diff
@@ -19,18 +19,30 @@
...
 
+DSN_DEFINE_int32("pegasus.server",
+                 coarse_data_variance_threshold,
+                 3,
+                 "the threshold of variance calculate to find the outliers");
+
+DSN_DEFINE_int32("pegasus.server",
+                 data_capture_hash_bucket_num,
+                 37,
+                 "the number of data capture hash buckets");
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:38]: you can add `DSN_DEFINE_validator` to make sure the `data_capture_hash_bucket_num` must be greater than 0

File: src/server/hotkey_collector.cpp [hunk 10]
```diff
@@ -32,11 +32,26 @@ DSN_DEFINE_int32("pegasus.server",
...
                  "the number of data capture hash buckets");
 
+DSN_DEFINE_validator(data_capture_hash_bucket_num, [](int32_t bucket_num) -> bool {
+    if (bucket_num < 3) {
+        return false;
+    }
+    // data_capture_hash_bucket_num should be a prime number
+    for (int i = 2; i <= bucket_num / i; i++) {
+        if (bucket_num % i == 0)
+            return false;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:50]:
```suggestion
        if (bucket_num % i == 0) {
            return false;
        }
```

File: src/server/hotkey_collector.cpp [hunk 11]
```diff
@@ -164,43 +179,49 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
         buckets[i] = _hash_buckets[i].load();
         _hash_buckets[i].store(0);
     }
-    result = internal_analysis_method(buckets, FLAGS_coarse_data_variance_threshold);
+    outlier_detection detection(buckets, FLAGS_coarse_data_variance_threshold);
+    int hotindex = -1;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:183]:
```suggestion
    int hot_index = -1;
```

File: src/server/hotkey_collector.cpp [hunk 12]
```diff
@@ -19,18 +19,90 @@
...
+    } else {
+        hot_index = -1;
+        return false;
+    }
+}
+
+static int get_bucket_id(dsn::string_view data)
+{
+    size_t hash_value = boost::hash_range(data.begin(), data.end());
+    return static_cast<int>(hash_value % FLAGS_data_capture_hash_bucket_num);
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:97]: Calculate hash value of every key may be wastly, is it possible to use https://github.com/XiaoMi/rdsn/blob/master/src/dsn.layer2.thrift#L92 directly?

Comment (Smityz) [reply src/server/hotkey_collector.cpp:97]: In the current design, it can't get this info

**Review by acelyc111 — DISMISSED** [review]

File: src/server/hotkey_collector.cpp (hunk 12 above)
Comment (acelyc111) [reply src/server/hotkey_collector.cpp:98]: Remember to improve it https://github.com/apache/incubator-pegasus/issues/628

**Review by acelyc111 — DISMISSED** [review]

File: src/server/hotkey_collector.cpp [hunk 13]
```diff
@@ -19,25 +19,98 @@
...
+
 DSN_DEFINE_int32(
     "pegasus.server",
     max_seconds_to_detect_hotkey,
     150,
     "the max time (in seconds) allowed to capture hotkey, will stop if hotkey's not found");
 
+// 68–95–99.7 rule, same algorithm as hotspot_partition_calculator::stat_histories_analyse
+static bool
+find_outlier_index(const std::vector<uint64_t> &captured_keys, int threshold, int &hot_index)
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:65]: The function parameter `threshold` is not needed, you can use `FLAGS_coarse_data_variance_threshold` directly.

Comment (hycdong) [inline src/server/hotkey_collector.cpp:65]: `captured_keys` is a copy of `_hash_bucket`, it seems that you wouldn't update the value of `_hash_buckets`, why would you copy another vector here?

File: src/server/hotkey_collector.cpp [hunk 14]
```diff
@@ -19,25 +19,98 @@
...
+    // TODO: (Tangyanzhao) increase a judgment of table_captured_key_sum
+    double captured_keys_avg_count =
+        (table_captured_key_sum - captured_keys[hot_index]) / (data_size - 1);
+    double standard_deviation = 0;
+    for (int i = 0; i < data_size; i++) {
+        if (i != hot_index) {
+            standard_deviation += pow((captured_keys[i] - captured_keys_avg_count), 2);
+        }
+    }
+    standard_deviation = sqrt(standard_deviation / (data_size - 2));
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:89]: Why use `data_size - 2` rather than `data_size - 1`?

File: src/server/hotkey_collector.cpp [hunk 15]
```diff
@@ -19,25 +19,98 @@
...
+    double hot_point = (hot_value - captured_keys_avg_count) / standard_deviation;
+    detect_hotkey_result result;
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:91]: What is the usage of `result` here? It seems is useless.

File: src/server/hotkey_collector.cpp (hunk 13 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:65]: It is a public function actually, the `fine_analyse` will use this function also

Comment (Smityz) [reply src/server/hotkey_collector.cpp:65]: `_hash_bucket` is dynamic changed, because in other threads it is always capturing data.
in other words, it has to keep `captured_keys` const when analyzing, otherwise, the result is incorrect.

File: src/server/hotkey_collector.cpp (hunk 14 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:89]: image
In the wiki's desc, the denominator should be `N-1`, so except for the data to be tested, it should be `data_size - 2`
but in some implements of this algorithm, it can also be `N`, that is `data_size`

File: src/server/hotkey_collector.cpp (hunk 15 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:91]: On the one hand, it can pass the `detect_hotkey_result::coarse_bucket_index` to `fine_collector`
on the other hand, it can be a snapshot to validate the final result

Comment (hycdong) [reply src/server/hotkey_collector.cpp:91]: I don't understand your explanation, `result` is the local variety of this function, your explanation seems to describe `_result` defined in `hotkey_collector`.

File: src/server/hotkey_collector.cpp (hunk 13 above)
Comment (hycdong) [reply src/server/hotkey_collector.cpp:65]: Okay, I have an another question now. If `_hash_buckets` will be updated in different threads, why don't you add a lock when read or write it?

File: src/server/hotkey_collector.cpp (hunk 15 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:91]: I will add more variable in `detect_hotkey_result`

File: src/server/hotkey_collector.cpp [hunk 16]
```diff
@@ -19,25 +19,97 @@
...
 #include <dsn/utility/flags.h>
 
 namespace pegasus {
 namespace server {
 
+DSN_DEFINE_int32("pegasus.server",
+                 coarse_data_variance_threshold,
+                 3,
+                 "the threshold of variance calculate to find the outliers");
+
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:35]: remove the blank line

**Review by hycdong — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]
